cpl_float.h needs NOMINMAX

### What is the bug?

Noticed in building the vcpkg port of PDAL while update to GDAL to the 3.11.0 release:

`cpl_float.h` defines `max` and `min` member functions for `Float16`. 

https://github.com/OSGeo/gdal/blob/96c33d68f71971d982b3b1d5e3143d4cd72b9f44/port/cpl_float.h#L532-L542

This code fails to compile under MSVC with the macro definitions from windows.h. Example:

~~~
FAILED: CMakeFiles/pdalcpp.dir/io/OGRWriter.cpp.obj 
C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1443~1.348\bin\Hostx64\x64\cl.exe   /TP -DH3_PREFIX="" -DPDAL_DLL_EXPORT=1 -DWIN32_LEAN_AND_MEAN -DWINDOWS -D_CRT_NONSTDC_NO_WARNING -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_USE_MATH_DEFINES -D_XKEYCHECK_H=1 -Dpdalcpp_EXPORTS -ID:\b\pdal\src\2.8.4-dc403abb10.clean -ID:\b\pdal\x64-windows-release-rel\include -ID:\b\pdal\src\2.8.4-dc403abb10.clean\vendor -ID:\b\pdal\src\2.8.4-dc403abb10.clean\vendor\nlohmann -ID:\installed\x64-windows-release\include\libxml2 -external:ID:\b\pdal\src\2.8.4-dc403abb10.clean\vendor\kazhdan -external:ID:\installed\x64-windows-release\include -external:ID:\installed\x64-windows-release\include\eigen3 -external:ID:\installed\x64-windows-release\include\utf8cpp -external:W0 /nologo /DWIN32 /D_WINDOWS /utf-8 /GR /EHsc /MP  /MD /O2 /Oi /Gy /DNDEBUG /Z7  -std:c++17 /wd4068 /wd4290 /wd4800 /wd4267 /wd4250 /wd4251 /EHsc /wd4996 /MP32 /showIncludes /FoCMakeFiles\pdalcpp.dir\io\OGRWriter.cpp.obj /FdCMakeFiles\pdalcpp.dir\ /FS -c D:\b\pdal\src\2.8.4-dc403abb10.clean\io\OGRWriter.cpp
D:\installed\x64-windows-release\include\cpl_float.h(532): error C2059: syntax error: '('
D:\installed\x64-windows-release\include\cpl_float.h(532): error C2143: syntax error: missing ',' before '?'
D:\installed\x64-windows-release\include\cpl_float.h(532): error C2535: 'cpl::Float16::Float16(void)': member function already defined or declared
D:\installed\x64-windows-release\include\cpl_float.h(294): note: see declaration of 'cpl::Float16::Float16'
~~~



### Steps to reproduce the issue

From WIP at https://github.com/microsoft/vcpkg/pull/45153.

Generalized, but not tested:
~~~
#include <windows.h>
#include <cpl_float.h>
~~~

### Versions and provenance

GDAL 3.11.0

(New code in this version, https://github.com/OSGeo/gdal/commit/ce7a57ae17a7c30f1bd7108ed9e220fe149fd55e)

### Additional context

I know the macros can be disabled with `NOMINMAX`. But it is not clear how to apply this in a general way without other unintented side effects. 

(FTR for the particular vcpkg pdal port problem, I apply `set_source_files_properties("${PROJECT_SOURCE_DIR}/io/OGRWriter.cpp" PROPERTIES COMPILE_DEFINITIONS NOMINMAX)` via `CMAKE_PROJECT_INCLUDE` script. I don't know if it it applies to *pristine* PDAL, and I don't have a MSVC machine for investigation. CC @hobu)

--- a/apps/gdalalg_raster_calc.cpp
+++ b/apps/gdalalg_raster_calc.cpp
@@ -15,6 +15,7 @@
 #include "../frmts/vrt/gdal_vrt.h"
 #include "../frmts/vrt/vrtdataset.h"
 
+#include "cpl_float.h"
 #include "gdal_priv.h"
 #include "gdal_utils.h"
 

--- a/autotest/cpp/test_gdal_algorithm.cpp
+++ b/autotest/cpp/test_gdal_algorithm.cpp
@@ -23,6 +23,7 @@
 
 #include "gtest_include.h"
 
+#include <algorithm>
 #include <limits>
 
 namespace test_gdal_algorithm

--- a/frmts/vrt/pixelfunctions.cpp
+++ b/frmts/vrt/pixelfunctions.cpp
@@ -16,6 +16,7 @@
 #include "vrtdataset.h"
 #include "vrtexpression.h"
 #include "vrtreclassifier.h"
+#include "cpl_float.h"
 
 #include <limits>
 

--- a/frmts/vrt/vrtprocesseddatasetfunctions.cpp
+++ b/frmts/vrt/vrtprocesseddatasetfunctions.cpp
@@ -10,6 +10,7 @@
  * SPDX-License-Identifier: MIT
  ****************************************************************************/
 
+#include "cpl_float.h"
 #include "cpl_minixml.h"
 #include "cpl_string.h"
 #include "vrtdataset.h"

--- a/frmts/zarr/zarr_v2_array.cpp
+++ b/frmts/zarr/zarr_v2_array.cpp
@@ -10,6 +10,7 @@
  * SPDX-License-Identifier: MIT
  ****************************************************************************/
 
+#include "cpl_float.h"
 #include "cpl_vsi_virtual.h"
 #include "gdal_thread_pool.h"
 #include "zarr.h"

--- a/frmts/zarr/zarr_v3_array.cpp
+++ b/frmts/zarr/zarr_v3_array.cpp
@@ -10,6 +10,7 @@
  * SPDX-License-Identifier: MIT
  ****************************************************************************/
 
+#include "cpl_float.h"
 #include "cpl_vsi_virtual.h"
 #include "gdal_thread_pool.h"
 #include "zarr.h"

--- a/gcore/gdal_priv.h
+++ b/gcore/gdal_priv.h
@@ -47,7 +47,6 @@ class GDALAlgorithm;
 #include "gdalsubdatasetinfo.h"
 #include "cpl_vsi.h"
 #include "cpl_conv.h"
-#include "cpl_float.h"
 #include "cpl_string.h"
 #include "cpl_minixml.h"
 #include "cpl_multiproc.h"
@@ -3748,10 +3747,10 @@ class CPL_DLL GDALMDArray : virtual public GDALAbstractMDArray,
     Transpose(const std::vector<int> &anMapNewAxisToOldAxis) const;
 
     std::shared_ptr<GDALMDArray> GetUnscaled(
-        double dfOverriddenScale = cpl::NumericLimits<double>::quiet_NaN(),
-        double dfOverriddenOffset = cpl::NumericLimits<double>::quiet_NaN(),
+        double dfOverriddenScale = std::numeric_limits<double>::quiet_NaN(),
+        double dfOverriddenOffset = std::numeric_limits<double>::quiet_NaN(),
         double dfOverriddenDstNodata =
-            cpl::NumericLimits<double>::quiet_NaN()) const;
+            std::numeric_limits<double>::quiet_NaN()) const;
 
     virtual std::shared_ptr<GDALMDArray>
     GetMask(CSLConstList papszOptions) const;
@@ -4627,29 +4626,21 @@ GDALDataset *GDALCreateOverviewDataset(GDALDataset *poDS, int nOvrLevel,
 // TODO: The expression `abs(fVal1 + fVal2)` looks strange; is this a bug?
 // Should this be `abs(fVal1) + abs(fVal2)` instead?
 
-inline bool ARE_REAL_EQUAL(GFloat16 dfVal1, GFloat16 dfVal2, int ulp = 2)
-{
-    using std::abs;
-    return dfVal1 == dfVal2 || /* Should cover infinity */
-           abs(dfVal1 - dfVal2) < cpl::NumericLimits<GFloat16>::epsilon() *
-                                      abs(dfVal1 + dfVal2) * ulp;
-}
-
 inline bool ARE_REAL_EQUAL(float fVal1, float fVal2, int ulp = 2)
 {
     using std::abs;
     return fVal1 == fVal2 || /* Should cover infinity */
            abs(fVal1 - fVal2) <
-               cpl::NumericLimits<float>::epsilon() * abs(fVal1 + fVal2) * ulp;
+               std::numeric_limits<float>::epsilon() * abs(fVal1 + fVal2) * ulp;
 }
 
-// We are using `cpl::NumericLimits<float>::epsilon()` for backward
+// We are using `std::numeric_limits<float>::epsilon()` for backward
 // compatibility
 inline bool ARE_REAL_EQUAL(double dfVal1, double dfVal2, int ulp = 2)
 {
     using std::abs;
     return dfVal1 == dfVal2 || /* Should cover infinity */
-           abs(dfVal1 - dfVal2) < cpl::NumericLimits<float>::epsilon() *
+           abs(dfVal1 - dfVal2) < std::numeric_limits<float>::epsilon() *
                                       abs(dfVal1 + dfVal2) * ulp;
 }
 

--- a/gcore/gdal_typetraits.h
+++ b/gcore/gdal_typetraits.h
@@ -27,16 +27,7 @@
 #define GDAL_TYPETRAITS_H_INCLUDED
 
 #include "gdal_priv.h"
-
-// NOTE: below GDAL_ENABLE_FLOAT16 is not guaranteed to be stable and is
-// mostly for Esri internal needs for now. Might be revisited if/once RFC 100
-// (https://github.com/OSGeo/gdal/pull/10146) is adopted.
-#ifdef GDAL_ENABLE_FLOAT16
-#if defined(__GNUC__) || defined(__clang__)
-#define __STDC_WANT_IEC_60559_TYPES_EXT__
-#include <float.h>  // Also brings in GFloat16
-#endif
-#endif
+#include "cpl_float.h"
 
 #include <complex>
 

--- a/gcore/gdalcachedpixelaccessor.h
+++ b/gcore/gdalcachedpixelaccessor.h
@@ -15,6 +15,7 @@
 
 #include "gdal_priv.h"
 #include "cpl_error.h"
+#include "cpl_float.h"
 
 #include <algorithm>
 #include <array>

--- a/gcore/gdalrasterband.cpp
+++ b/gcore/gdalrasterband.cpp
@@ -4089,6 +4089,18 @@ struct GDALNoDataValues
 };
 
 /************************************************************************/
+/*                            ARE_REAL_EQUAL()                          */
+/************************************************************************/
+
+inline bool ARE_REAL_EQUAL(GFloat16 dfVal1, GFloat16 dfVal2, int ulp = 2)
+{
+    using std::abs;
+    return dfVal1 == dfVal2 || /* Should cover infinity */
+           abs(dfVal1 - dfVal2) < cpl::NumericLimits<GFloat16>::epsilon() *
+                                      abs(dfVal1 + dfVal2) * ulp;
+}
+
+/************************************************************************/
 /*                            GetHistogram()                            */
 /************************************************************************/
 

